torch/csrc/distributed/rpc/request_callback_no_python.cpp

```diff
@@ -583,25 +502,48 @@ void RequestCallbackNoPython::processRpc(
             wrappedMsgType,
             messageId,
             wrappedRpcResponseFuture);
-      }
-      wrappedRpcResponseFuture->addCallback([wrappedRpcResponseFuture,
+
+        auto tid = std::this_thread::get_id();
+        wrappedRpcResponseFuture->addCallback(
+            at::wrapPropagateTLSState<void>([wrappedRpcResponseFuture,
                                              responseFuture,
-                                             profiledEvents =
-                                                 std::move(profiledEvents),
-                                             profilingKeyId] {
-        if (wrappedRpcResponseFuture->hasError()) {
-          // Propagate error
-          responseFuture->setError(wrappedRpcResponseFuture->error()->what());
-        } else {
-          auto rpcWithProfilingResp = std::make_unique<RpcWithProfilingResp>(
-              MessageType::RUN_WITH_PROFILING_RESP,
-              std::move(*wrappedRpcResponseFuture).moveValue(),
-              profiledEvents,
-              profilingKeyId);
-          responseFuture->markCompleted(
-              std::move(*rpcWithProfilingResp).toMessage());
-        }
-      });
+                                             profilingKeyId,
+                                             profilingConfig,
+                                             tid] {
+              std::vector<torch::autograd::profiler::Event> profiledEvents;
+              // Defer consolidation of profiler events until async work has
+              // completed (such as async UDF)
+
+              TORCH_INTERNAL_ASSERT(
+                  torch::autograd::profiler::profilerEnabled(),
+                  "Expected profiler to be enabled!");
+
+              // Only clean up TLS states of profiler if we are disabling on
+              // the main thread.
+              bool shouldCleanUpTLSStates = (std::this_thread::get_id() == tid);
```

Comment:
could you please elaborate more why we don't need clean up TLS on child thread?

Reply:
Per some discussions with @ilia-cher, the way I understand it is that child threads with their thread-local states propagated over by parent threads will pop their thread local states after async task is finished executing.

Since the profiler wasn't explicitly enabled in this thread (rather carried over by `at::wrapPropagateTLSState`) we shouldn't explicitly clean up the TL states of profiler either (which is what disableProfiler() would do by default). Those states will be clean up after the guard in `wrapPropagateTLSState` exits.

Reply:
I'm assuming `shouldCleanUpTLSStates` is only True when `wrappedRpcResponseFuture` has already completed and is running inline as a result? If so, should we do this cleanup logic after calling `addCallback` so that it runs on the main thread and cleans up the state accordingly? We could run the cleanup logic only if we know the future has completed.

It's messy to have code like this which is conditional on thread ids.

Reply:
In addition to the case that you mentioned, I believe it's also possible to encounter this when running with single-threaded server where all callbacks will run on the same thread. Also, with a thread pool I think there also might be a chance where we run the callback on the same thread that created the future (it would happen if the thread in the thread pool that called `markCompleted()` happened to be the same). 

Although, I think we can remove the thread_id conditioning and just never clean up TLS state in the callback. The reason is, in the case where we run inline/on the main thread, we already wrap this callback with an extra layer of TLSState, so we would not want to clean up this extra state (it would be automatically cleaned by the `ThreadLocalStateGuard` destructor).

Reply:
Looks like this works, also added a test in `jit/test_misc.cpp` which should simulate this issue much more cleanly.

Reply:
> The reason is, in the case where we run inline/on the main thread, we already wrap this callback with an extra layer of TLSState, so we would not want to clean up this extra state (it would be automatically cleaned by the ThreadLocalStateGuard destructor). 


Is this true for all functions run by thread from the pool?

Reply:
It's actually only true for specific callbacks such as this one: ones that are wrapped with `at::wrapPropagateTLSState` (or otherwise use `ThreadLocalStateGuard`, but `wrapPropagateTLSState` is the recommended API for continuation callbacks). It won't carry over thread local state implicitly for an arbitrary function ran by the thread pool.